Allow upcasting

captain-yoshi opened the pull request:
Unit tests with desired behavior from #943. 


### Any Tests Failures
```bash
/home/captain-yoshi/ws/ros/mimik_ws/src/BehaviorTree.CPP/tests/gtest_any.cpp:271: Failure
Expected: auto res = a.cast<std::shared_ptr<Greeter>>() doesn't throw an exception.
  Actual: it throws.
/home/captain-yoshi/ws/ros/mimik_ws/src/BehaviorTree.CPP/tests/gtest_any.cpp:273: Failure
Value of: a.castPtr<std::shared_ptr<Greeter>>()
  Actual: false
Expected: true
```

### Blackboard Test Failure
```bash
C++ exception with description "The creation of the tree failed because the port [hello_greeter] was initially created with type [std::shared_ptr<HelloGreeter>] and, later type [std::shared_ptr<Greeter>] was used somewhere else." thrown in the test body.
```

captain-yoshi commented:
*There maybe a better way to allow type casting of a polymorphic class.*

​This pull request introduces support for upcasting and downcasting `std::shared_ptr` instances (must be polymorphic and has their base registered) within the `Any` class, facilitating polymorphic type handling. By leveraging type traits, the `Any` class can now perform safe and efficient conversions between shared pointers of related types.
- Introduced `any_cast_base<T>` trait to register base types for polymorphic casting.
- Added `is_shared_ptr` and `IsPolymorphicSharedPtr` traits to detect eligible types.
- Modified `Any` to store shared_ptr as its registered base class if available.
- Enhanced castPtr() and tryCast() to support safe downcasting from stored base class to derived class using static or dynamic pointer casting.
- Ensures shared_ptr<Derived> can be stored and retrieved as shared_ptr<Base> when registered.

So users who wants this behavior needs to register their classes to it's base class. I don't think we can get away without this process...
```cpp
// Register cast base type for self to allow direct cast, otherwise defaults to a dynamic cast
template <>
struct BT::any_cast_base<Greeter>
{
  using type = Greeter;
};

// Register cast base type for HelloGreeter
template <>
struct BT::any_cast_base<HelloGreeter>
{
  using type = Greeter;
};
```
Optionally we could add a macro to simplify this process (not in this PR):
```cpp
#define BT_REGISTER_BASE_TYPE(Derived, Base)    \
  template <>                                   \
  struct BT::any_cast_base<Derived>             \
  {                                             \
    using type = Base;                          \
  };
  
BT_REGISTER_BASE_TYPE(Greeter, Greeter)
BT_REGISTER_BASE_TYPE(HelloGreeter, Greeter) 
```

It's a little bit more permissive, then what I intended initially. You can now downcast to something that is invalid (must still be a valid class of the stored base class), e.g. the blackboard will not see a mismatch, but it will fail when trying to retrieve it in a BT::Node. 

Regarding the previous point, should we try to downcast early, in the blackboard to stop the tree from running early if it is invalid ? Instead of running the tree and let it fail down the line ?

----
The Sonarcube check fails, but it does not seem related to this PR.

captain-yoshi commented:
The `castPtr` function is a little bit tricky when downcasting...

I would prefer solution 2 or 3.

### Solution 1
Returns a `T::element_type* castPtr()` when T is a shared pointer with a polymorphic class and a registered base. For everything else T* is returned.

| Pros          | Cons                                                                                                                   |
|---------------|------------------------------------------------------------------------------------------------------------------------|
| No complexity | 2 different return types depending on the type.<br>Cannot retrieve the shared pointer, only access to the raw pointer. |

*Implemented in 3021ff0*

### Solution 2
Clean API: `T* castPtr()` for all types.
| Pros       | Cons                                                                                                                                                                       |
|------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Returns T* | Need to store a std_shared\<void> member to return a pointer to a Derived class.<br>Lifetime of pointer not garanteed*, another call to castPtr may change this cached value. <br><br>*If getLockedPortContent is used, there is no problem.|

*Implemented in d11512a*

### Solution 3
Seperate castPtr function when T is a shared pointer with a polymorphic class and a registered base, e.g. `castShared`.
| Pros          | Cons                                                                    |
|---------------|-------------------------------------------------------------------------|
| No complexity | User has 2 different functions depending on what is stored in the _any. |

*Not implemented*

facontidavide commented:
There is a conflict that need to be resolved. Also, correct me if i am wrong, but isn't this change allowing two ports with derived classes to connect?

For instance, given **Cat <- Animal** and **Dog <- Animal**, I would be allowed to connect a Cat to a Dog...

captain-yoshi commented:
> For instance, given Cat <- Animal and Dog <- Animal, I would be allowed to connect a Cat to a Dog...

At the moment yes. If we want to make it fail when creating a Tree node from xml, we would need to store the type id of the base class and the original class, because at this stage, I think the any values are still empty.
<p align="center">
<img src="https://github.com/user-attachments/assets/7eca02c1-a60d-4af1-8ec1-ddff9bf0cc32" width="250">
</p>

But trying to have a SphinxCat into a registered Cat port would also fail because only the the SphinxCat and Animal typeid would be stored when registering SphinxCat. We could fix this by storing a vector (or limit with a max depth) of typeid's from current class to root base class to enable this case.

I would also prefer to make it fail early when trying to connect incompatible types. What wold be the way to go ? Only allow a depth of 2 base classes, allow a max range of depth or allow any depth ? Or something else entirely...

captain-yoshi commented:
This succeeds:
```cpp
std::string xml_txt = R"(
<root BTCPP_format="4" >
  <BehaviorTree ID="Main">
    <Sequence>

      <CreateCat out_cat="{cat}" />
      <CreateAnimal in_animal="{cat}" />
    </Sequence>
  </BehaviorTree>
</root>)";

BehaviorTreeFactory factory;
factory.registerNodeType<CreateAnimal>("CreateAnimal");
factory.registerNodeType<CreateCat>("CreateCat");
factory.registerNodeType<CreateSphynxCat>("CreateSphynxCat");
factory.registerNodeType<CreateDog>("CreateDog");

auto tree = factory.createTreeFromText(xml_txt);

ASSERT_EQ(status, NodeStatus::SUCCESS);
```
and also this succeeds:
```cpp
std::string xml_txt = R"(
<root BTCPP_format="4" >
  <BehaviorTree ID="Main">
    <Sequence>
      <CreateCat out_cat="{cat}" />
      <CreateSphynxCat in_sphynx="{cat}" />
    </Sequence>
  </BehaviorTree>
</root>)";

BehaviorTreeFactory factory;
factory.registerNodeType<CreateAnimal>("CreateAnimal");
factory.registerNodeType<CreateCat>("CreateCat");
factory.registerNodeType<CreateSphynxCat>("CreateSphynxCat");
factory.registerNodeType<CreateDog>("CreateDog");

ASSERT_ANY_THROW(auto tree = factory.createTreeFromText(xml_txt));
```

but this should throw at creation but succeeds.
```cpp
std::string xml_txt = R"(
<root BTCPP_format="4" >
  <BehaviorTree ID="Main">
    <Sequence>

      <CreateCat out_cat="{cat}" />
      <CreateAnimal out_animal="{cat}" />
      <CreateCat out_cat="{cat}" />
    </Sequence>
  </BehaviorTree>
</root>)";

BehaviorTreeFactory factory;
factory.registerNodeType<CreateAnimal>("CreateAnimal");
factory.registerNodeType<CreateCat>("CreateCat");
factory.registerNodeType<CreateSphynxCat>("CreateSphynxCat");
factory.registerNodeType<CreateDog>("CreateDog");

ASSERT_ANY_THROW(auto tree = factory.createTreeFromText(xml_txt));

// Throws further down in the Blackboard::set of the CreateAnimal
// NodeStatus status = tree.tickWhileRunning();
```

Upcasting from Input port is correct, but upcasting to Output ports must not. But we currently don't have the port direction info, so I cannot preempt early in the createNodeFromXml function:
```cpp
for(const auto& [port_name, port_info] : manifest->ports)
{
  auto remap_it = port_remap.find(port_name);
  if(remap_it == port_remap.end())
  {
    continue;
  }
  StringView remapped_port = remap_it->second;

  if(auto param_res = TreeNode::getRemappedKey(port_name, remapped_port))
  {
    // port_key will contain the key to find the entry in the blackboard
    const auto port_key = static_cast<std::string>(param_res.value());

    // if the entry already exists, check that the type is the same
    if(auto prev_info = blackboard->entryInfo(port_key))
    {
      // Check consistency of types.
      bool port_type_mismatch =
          (prev_info->isStronglyTyped() && port_info.isStronglyTyped() &&
           prev_info->type() != port_info.type());

      // *********************************************************
      // Only allow the next mismatch if the port_info is an input or bidirectionnal ?

      // allow mismatch if the previous type appears in the base_chain
      // of the current port (i.e., valid polymorphic upcast)
      if(port_type_mismatch && !prev_info->baseChain().empty() &&
         !port_info.baseChain().empty() &&
         prev_info->isConvertibleFrom(port_info.baseChain()[0]))
      {
        port_type_mismatch = false;
      }

      // special case related to convertFromString
      bool const string_input = (prev_info->type() == typeid(std::string));

      if(port_type_mismatch && !string_input)
      {
        blackboard->debugMessage();

        throw RuntimeError("The creation of the tree failed because the port [",
                           port_key, "] was initially created with type [",
                           demangle(prev_info->type()), "] and, later type [",
                           demangle(port_info.type()), "] was used somewhere else.");
      }
    }
    else
    {
      // not found, insert for the first time.
      blackboard->createEntry(port_key, port_info);
    }
  }
}
```  
Is the port direction really needed ? If yes how do you propose to have access to this information ?

captain-yoshi commented:
I have now replaced the Greeter hierarchy with the Animal one (easier too understand).

I added some new tests. **Cat <- Animal** and **Dog <- Animal** now throws when constructing the TreeNode.

----
I think the last thing missing is the serialization of the TreeNodeModel in XML ? I think we would need to store the chain base, e.g. :
```xml
<output_port name="collision_object"
             type="std::shared_ptr&lt;Sphinx&lt;std::allocator&lt;void&gt; &gt; &gt;"
             base_chain="Sphinx;Cat;Animal" />
```

Do we need something for deserializing the base_chain ? Is this even needed, I got the impression that we always need to register the node type in the BehaviorTreeFactory...

facontidavide commented:
thanks for working on this. 

the proposed solution seems to work but it adds some (necessary) complexity and I need to take my time (something that right now I don't have in abundance, to think about alternatives... if they exist.

Please be patient, i will look at this in the next couple of weeks.
